### ISSUE TITLE
A compile error in a dependent package causes hPutChar: invalid argument (invalid character) 

### ISSUE THREAD

**Issue by fumieval** [commercialhaskell/stack#422]
After #406, another problem happens. `decodeUtf8With lenientDecode` seems to be unable to decode ghc's output well.

```
prelude-kai-0.1.0.0: build
Progress: 1/2
--  While building package prelude-kai-0.1.0.0 using:
      E:\\minghc-7.10.1-x86_64\\ghc-7.10.1\\bin\\runhaskell.exe -package=Cabal-1
.22.2.0 -clear-package-db -global-package-db -package-db=E:\hs\.stack\snapshots\
x86_64-windows\nightly-2015-06-25\7.10.1\pkgdb\ C:\Users\Fumiaki\AppData\Local\T
emp\stack648\Setup.hs --builddir=.stack-work\dist\x86_64-windows\Cabal-1.22.2.0\
 build
    Process exited with code: ExitFailure 1
    Logs have been written to: "E:\\hs\\Haswerk\\.stack-work\\logs\\prelude-kai-
0.1.0.0.log"

    Building prelude-kai-0.1.0.0...
    Preprocessing library prelude-kai-0.1.0.0...
    [1 of 1] Compiling Prelude.Kai      ( src\Prelude\Kai.hs, .stack-work\dist\x
86_64-windows\Cabal-1.22.2.0\build\Prelude\Kai.o )

    src\Prelude\Kai.hs:200:16:
        No instance for (Num [Char]) arising from the literal stack: <stdout>: h
PutChar: invalid argument (invalid character)
```

**Comment by snoyberg** [comment]
Good catch. I've pushed (what I believe to be) a fix to master, can you give it a shot?

**Comment by fumieval** [comment]
Fast! It now works. There are still some _mojibake_ though.

```
E:\hs\Haswerk>stack build
Warning: File listed in E:\GitHub\freetype2\freetype2.cabal file does not exist:
 objs/ftmodule.h
prelude-kai-0.1.0.0: build

--  While building package prelude-kai-0.1.0.0 using:
      E:\\minghc-7.10.1-x86_64\\ghc-7.10.1\\bin\\runhaskell.exe -package=Cabal-1
.22.2.0 -clear-package-db -global-package-db -package-db=E:\hs\.stack\snapshots\
x86_64-windows\nightly-2015-06-25\7.10.1\pkgdb\ C:\Users\Fumiaki\AppData\Local\T
emp\stack2832\Setup.hs --builddir=.stack-work\dist\x86_64-windows\Cabal-1.22.2.0
\ build
    Process exited with code: ExitFailure 1
    Logs have been written to: "E:\\hs\\Haswerk\\.stack-work\\logs\\prelude-kai-
0.1.0.0.log"

    Building prelude-kai-0.1.0.0...
    Preprocessing library prelude-kai-0.1.0.0...
    [1 of 1] Compiling Prelude.Kai      ( src\Prelude\Kai.hs, .stack-work\dist\x
86_64-windows\Cabal-1.22.2.0\build\Prelude\Kai.o )

    src\Prelude\Kai.hs:200:8:
        Not in scope: data constructor ・ｽeBlah・ｽf
        Perhaps you meant variable ・ｽeblah・ｽf (line 200)
    ghc: warning: _tzset from msvcrt is linked instead of __imp__tzset
Pr
E:\hs\Haswerk>

Not in scope: data constructor ・ｽeBool・ｽf
```

**Comment by snoyberg** [comment]
What do you see if you run:

    stack exec cat .stack-work/logs/prelude-kai-0.1.0.0.log

**Comment by fumieval** [comment]
Oddly enough,
```
Building prelude-kai-0.1.0.0...
Preprocessing library prelude-kai-0.1.0.0...
[1 of 1] Compiling Prelude.Kai      ( src\Prelude\Kai.hs, .stack-work\dist\x86_6
4-windows\Cabal-1.22.2.0\build\Prelude\Kai.o )
src\Prelude\Kai.hs:200:8:
    Not in scope: data constructor ▒eBlah▒f
    Perhaps you meant variable ▒eblah▒f (line 200)
ghc: warning: _tzset from msvcrt is linked instead of __imp__tzset
```

And Here's the content of `.stack-work/logs/prelude-kai-0.1.0.0.log` directly copied from the editor (CP932):

```
Building prelude-kai-0.1.0.0...
Preprocessing library prelude-kai-0.1.0.0...
[1 of 1] Compiling Prelude.Kai      ( src\Prelude\Kai.hs, .stack-work\dist\x86_64-windows\Cabal-1.22.2.0\build\Prelude\Kai.o )

src\Prelude\Kai.hs:200:8:
    Not in scope: data constructor ‘Blah’
    Perhaps you meant variable ‘blah’ (line 200)
ghc: warning: _tzset from msvcrt is linked instead of __imp__tzset
```

**Comment by snoyberg** [comment]
That's what I expected: it seems like your terminal itself doesn't support the Unicode characters GHC is outputting. Do you see something like that when using GHC directly (e.g., `stack ghc ...`)?

**Comment by fumieval** [comment]
ghc seemed to spit CP932, according to my terminal. `stack ghc` is fine.

**Comment by snoyberg** [comment]
OK, now I think I get it, or at least part of it: when we decode the .log file, we're assuming UTF-8, which isn't correct. GHC may output a different encoding. Our two options would be (1) force GHC to use UTF-8 by setting an environment variable, or (2) reading using Data.Text.IO instead, which will follow the same character encoding logic as GHC.

My concern, though, is that we'll end up in a situation where the IO system throws an exception, like you ran into originally. The nice thing about our current way of doing things is that we know that won't happen, though (as you've discovered) it can get characters wrong in some cases.

I'm open to ideas on this one, I'm not sure what the best path forward is.

**Comment by snoyberg** [comment]
For now, I'm going to close this issue, as there doesn't seem to be a perfect way to implement this without introducing the possibility of some systems of getting an exception when printing the exception... which is clearly not what we want.
